scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -436,112 +444,237 @@
+          (if (contains? matched-job-uuids pending-job-uuid)
+            (if backfilling?
+              (recur scheduler-contents' true matched-head? fully-processed upgrade-backfill (conj backfill-jobs pending-job-uuid))
+              (recur scheduler-contents' false (or pending=head? matched-head?) (conj fully-processed pending-job-uuid) upgrade-backfill backfill-jobs))
+            (if-let [backfilled-ids (and (not backfilling?) (seq (ids-of-backfilled-instances pending-job)))]
+              (recur scheduler-contents' false (or pending=head? matched-head?) fully-processed (into upgrade-backfill backfilled-ids) backfill-jobs)
+              (recur scheduler-contents' true matched-head? fully-processed upgrade-backfill backfill-jobs))))
+        {:fully-processed fully-processed
+         :upgrade-backfill upgrade-backfill
+         :backfill-jobs backfill-jobs
+         :matched-head? matched-head?}))))
+
+(defn handle-resource-offers!
+  "Gets a list of offers from mesos. Decides what to do with them all--they should all
+   be accepted or rejected at the end of the function."
+  [conn driver ^TaskScheduler fenzo fid pending-jobs num-considerable offers-chan offers]
+  (log/debug "invoked handle-resource-offers!")
+  (let [offer-stash (atom nil)] ;; This is a way to ensure we never lose offers fenzo assigned if an errors occures in the middle of processing
+    (timers/time!
+      handle-resource-offer!-duration
+      (try
+        (loop [] ;; This loop is for compare-and-set! below
+          (let [scheduler-contents @pending-jobs
+                db (db conn)
+                _ (log/debug "There are" (count scheduler-contents) "pending jobs")
+                considerable (->> scheduler-contents
+                                  (filter (fn [job]
+                                            (job-allowed-to-start? db job)))
+                                  (take num-considerable))
+                _ (log/debug "We'll consider scheduling" (count considerable) "of those pending jobs (limited to " num-considerable " due to backdown)")
+                matches (match-offer-to-schedule fenzo considerable offers db fid)
+                _ (reset! offer-stash (doall (for [{:keys [leases]} matches
+                                                   lease leases]
+                                               (:offer lease))))
+                matched-jobs (for [match matches
+                                   ^TaskAssignmentResult task-result (:tasks match)
+                                   :let [task-request (.getRequest task-result)]]
+                               (:job task-request))
+                processed-matches (process-matches-for-backfill scheduler-contents (first considerable) matched-jobs)
+                new-scheduler-contents (remove (fn [pending-job]
+                                                 (contains? (:fully-processed processed-matches) (:job/uuid pending-job)))
```

Comment:
I think we should remove upgraded jobs as well. I don't think it matters but you will end up with a bunch of upgraded jobs at the front of scheduler contents.